packages/protocol/contracts/governance/DowntimeSlasherIntervals.sol

```diff
@@ -0,0 +1,409 @@
+pragma solidity ^0.5.3;
+
+import "openzeppelin-solidity/contracts/ownership/Ownable.sol";
+import "openzeppelin-solidity/contracts/math/SafeMath.sol";
+
+import "./SlasherUtil.sol";
+
+contract DowntimeSlasherIntervals is SlasherUtil {
+  using SafeMath for uint256;
+
+  // Intervals previously slashed
+  struct SlashedInterval {
+    // startBlock of the slashed interval
+    uint256 startBlock;
+    // endBlock of the slashed interval. Although this could be calculated, as the downtime interval
+    // could be changed, we store the calculated end to avoid future errors if the interval changes.
+    uint256 endBlock;
+  }
+
+  // For each address, associate each epoch with the last block that was slashed on that epoch
+  mapping(address => mapping(uint256 => SlashedInterval[])) epochSlashedIntervals;
+
+  // For each user a map of StartBlock to a map of EndBlock to an accumulated ParentSealBitmap for the Interval
+  // Maps start block -> end block -> bitmap
+  // Intervals are not allowed to cross epoch boundaries.
+  mapping(address => mapping(uint256 => mapping(uint256 => uint256))) private userIntervalValidationProof;
+
+  uint256 public slashableDowntime;
+  bool public oncePerEpoch;
+
+  event SlashableDowntimeSet(uint256 interval);
+  event OncePerEpochSet(bool oncePerEpoch);
+  event DowntimeSlashPerformed(address indexed validator, uint256 indexed startBlock);
+  event ProofOfIntervalValidationPerformed(
+    address indexed user,
```

Comment:
The actual event or just the use who did it?
I've added the event because it could be useful for the user to know which interval did he set in the pass, to eventually re use them, but we could remove them

Reply:
Just the user

Reply:
User removed